index.js

```diff
@@ -20,302 +36,589 @@ var prefs = Cc["@mozilla.org/preferences-service;1"].getService(Ci.nsIPrefServic
+            (doc !== undefined) && (doc !== null)) { // For Widgets
+          var win = getWidgetWindow(tickerId);
+          if (win) {
+            win.postMessage({
+              "type": "updateTickerModelPrice",
+              "id": tickerId,
+              "data": {
+                "price": price
               }
-              return;
-            }
-            price = price[jsonPath[i]];
-          }
-          if (DEBUG) {
-            console.log(TAG + " Price received and parsed: " + price);
+            }, "*");
+            adjustWidgetSize(tickerId);
           }
+        } else if (tickersFrame !== null) { // For Toolbar
           tickersFrame.postMessage({
             "type": "updateTickerModelPrice",
-            "id": id,
+            "id": tickerId,
             "data": {
               "price": price
             }
           }, tickersFrame.url);
+        } else {
+          dlog("Frame and Widget document are both empty for Widget " + tickerId);
         }
       }
-    }).get();
-  }
+    }
+  }).get();
+}
+
+
+
+// Methods to create tickers and manipulate them
+
+
+/**
+ * Create a Widget, attach listeners and trigger update on Widget Added to window
+ **/
+function createNewTickersWidget(tickerId) {
+  dlog("Creating widget for ticker " + tickerId);
+  var tickerTitle = tickers[tickerId].exchangeName + " " + 
+                      tickers[tickerId].currency + "/" + tickers[tickerId].baseCurrency
+  CustomizableUI.createWidget({
+    id:           tickerId + WIDGET_SUFFIX,
+    type:         "custom",
+    label:        tickerId,
+    removable:    true,
+    defaultArea:  CustomizableUI.AREA_NAVBAR,
+    /**
+     * Build widget iFrame content and attach listeners
+     **/
+    onBuild: function(aDocument) {
+      tickers[tickerId].doc = aDocument;
+      var node = aDocument.createElement("toolbaritem");
+      var props = {
+        id:           this.id,
+        title:        "Bitcoin Price Ticker " + tickerId,
+        align:        "center",
+        label:        tickerTitle,
+        width:        "15px",
+        height:       10,
+        tooltiptext:  tickerTitle,
+        class:        "chromeclass-toolbar-additional panel-wide-item"
+      };
+      for (var p in props) {
+        node.setAttribute(p, props[p])
+      }
 
-  function startAutoPriceUpdate(tickerId) {
-    if (tickers[tickerId].url && tickers[tickerId].jsonPath) {
-      var fetchURLDataWrapper = function() {
-        fetchURLData(tickerId, tickers[tickerId].url, tickers[tickerId].jsonPath);
+      var iframe = aDocument.createElement("iframe");
+      props = {
+        id:          tickerId + IFRAME_SUFFIX,
+        type:        "content",
+        src:         IFRAME_URL,
+        transparent: true
       };
-      fetchURLDataWrapper();
-      tickers[tickerId].timer = setInterval(fetchURLDataWrapper, 
-                                            (tickers[tickerId].updateInterval * 1000));
-    }
-  }
+      for (var p in props) {
```

Comment:
'p' is already defined.<br>The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.